plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php

```diff
@@ -297,14 +297,25 @@ private static function order_received( $order_id = 0 ) {
 			return;
 		}
 
-		$order_customer_id = $order->get_customer_id();
-
-		// For non-guest orders, require the user to be logged in before showing this page.
-		if ( $order_customer_id && get_current_user_id() !== $order_customer_id ) {
-			wc_get_template( 'checkout/order-received.php', array( 'order' => false ) );
-			wc_print_notice( esc_html__( 'Please log in to your account to view this order.', 'woocommerce' ), 'notice' );
-			woocommerce_login_form( array( 'redirect' => $order->get_checkout_order_received_url() ) );
-			return;
+		/**
+		 * Allow to guest users see the page order-received
+		 *
+		 * @param bool $order_received_verification_required If order-received verification is required.
+		 *
+		 * @since 8.4.0
+		 */
+		$order_received_verification_required = apply_filters( 'woocommerce_order_received_verification_required', true );
+
+		if($order_received_verification_required) {
+			$order_customer_id = $order->get_customer_id();
+	
+			// For non-guest orders, require the user to be logged in before showing this page.
+			if ( $order_customer_id && get_current_user_id() !== $order_customer_id ) {
+				wc_get_template( 'checkout/order-received.php', array( 'order' => false ) );
+				wc_print_notice( esc_html__( 'Please log in to your account to view this order.', 'woocommerce' ), 'notice' );
+				woocommerce_login_form( array( 'redirect' => $order->get_checkout_order_received_url() ) );
+				return;
+			}
```

Comment:
Incorporating the variable change from my last comment, we could do something like this here:

```suggestion
		$order_customer_id = $order->get_customer_id();
	
		// For non-guest orders, require the user to be logged in before showing this page.
		if ( $verify_known_shoppers && $order_customer_id && get_current_user_id() !== $order_customer_id ) {
			wc_get_template( 'checkout/order-received.php', array( 'order' => false ) );
			wc_print_notice( esc_html__( 'Please log in to your account to view this order.', 'woocommerce' ), 'notice' );
			woocommerce_login_form( array( 'redirect' => $order->get_checkout_order_received_url() ) );
			return;
		}
```

This also reduces the amount of nesting (arguably keeping things a little more readable). Alternatively, if we prefer to keep the nested `if` structure, we should correct the spacing to satisfy our coding standards.

Reply:
My initial idea on line 309 was do the verification of the filter before get the user information to improve performance and not get those informations if the filter are seted to false, but, since we have the order already loaded it may doesn't make a lot of difference; I also agree that with those changes the code is more readable.